core/src/main/java/org/apache/calcite/jdbc/SimpleCalciteSchema.java

```diff
@@ -67,31 +69,79 @@ public CalciteSchema add(String name, Schema schema) {
     return calciteSchema;
   }
 
+  private String caseInsensitiveLookup(Set<String> candidates, String name) {
+    // Exact string lookup
+    if (candidates.contains(name)) {
+      return name;
+    }
+    // Upper case string lookup
+    final String upperCaseName = name.toUpperCase(Locale.ROOT);
+    if (candidates.contains(upperCaseName)) {
+      return upperCaseName;
+    }
+    // Lower case string lookup
+    final String lowerCaseName = name.toLowerCase(Locale.ROOT);
+    if (candidates.contains(lowerCaseName)) {
+      return lowerCaseName;
+    }
+    // Fall through: Set iteration
+    for (String candidate: candidates) {
+      if (candidate.equalsIgnoreCase(name)) {
+        return candidate;
+      }
+    }
+    return null;
+  }
+
   protected CalciteSchema getImplicitSubSchema(String schemaName,
       boolean caseSensitive) {
     // Check implicit schemas.
-    Schema s = schema.getSubSchema(schemaName);
-    if (s != null) {
-      return new SimpleCalciteSchema(this, s, schemaName);
+    if (caseSensitive) {
+      Schema s = schema.getSubSchema(schemaName);
+      if (s != null) {
+        return new SimpleCalciteSchema(this, s, schemaName);
+      }
+    } else {
+      final String schemaName2 = caseInsensitiveLookup(
+          schema.getSubSchemaNames(), schemaName);
+      if (schemaName2 != null) {
+        return new SimpleCalciteSchema(this,
+            schema.getSubSchema(schemaName2), schemaName2);
```

Comment:
This looks like code duplication.
Can you please factor that out of `if` statement?

For instance:
1. determine schema name
2. return candidate

Reply:
Done